From 68149b669a26bac65177faf496000f7b433401dd Mon Sep 17 00:00:00 2001
From: Jeremy Allison <jra@samba.org>
Date: Wed, 16 Sep 2015 12:03:34 -0700
Subject: [PATCH] s3: smbd: Fix opening/creating :stream files on the root
 share directory.

BUG: https://bugzilla.samba.org/show_bug.cgi?id=11522

Signed-off-by: Jeremy Allison <jra@samba.org>
---
 source3/smbd/filename.c | 22 ++++++++++++++++++++++
 1 file changed, 22 insertions(+)

diff --git a/source3/smbd/filename.c b/source3/smbd/filename.c
index 6899d2a..422f53d 100644
--- a/source3/smbd/filename.c
+++ b/source3/smbd/filename.c
@@ -371,6 +371,28 @@ NTSTATUS unix_convert(TALLOC_CTX *ctx,
 			 */
 			*stream = '\0';
 			stream = tmp;
+
+			if (smb_fname->base_name[0] == '\0') {
+				/*
+				 * orig_name was just a stream name.
+				 * This is a stream on the root of
+				 * the share. Replace base_name with
+				 * a "."
+				 */
+				smb_fname->base_name =
+					talloc_strdup(smb_fname, ".");
+				if (smb_fname->base_name == NULL) {
+					status = NT_STATUS_NO_MEMORY;
+					goto err;
+				}
+				if (SMB_VFS_STAT(conn, smb_fname) != 0) {
+					status = map_nt_error_from_unix(errno);
+					goto err;
+				}
+				DEBUG(5, ("conversion finished \"\" -> %s\n",
+					  smb_fname->base_name));
+				goto done;
+			}
 		}
 	}
 
-- 
2.6.0.rc0.131.gf624c3d

